File: services/rag/app/services/attachment_worker.py
from __future__ import annotations

import threading
import uuid

from app.config import settings
from app.services.document_parser import AttachmentDocumentParser
from app.services.index_service import delete_document_chunks
from app.services.pgvector_store import PgVectorStore


class AttachmentParseWorker:
    """Consumes uploaded attachments in a process separate from text embeddings."""

    def __init__(self, store: PgVectorStore, parser: AttachmentDocumentParser):
        self.store, self.parser = store, parser
        self.worker_id = f"rag-attachment-parse-{uuid.uuid4().hex[:8]}"
        self.stop_event = threading.Event()
        self.thread: threading.Thread | None = None

    def start(self) -> None:
        if settings.attachment_parse_worker_enabled and self.thread is None:
            self.thread = threading.Thread(target=self.run, name="attachment-parse-worker", daemon=True)
            self.thread.start()

    def stop(self) -> None:
        self.stop_event.set()
        if self.thread:
            self.thread.join(timeout=10)
        self._heartbeat("stopped")

    def run(self) -> None:
        self._heartbeat("running")
        while not self.stop_event.is_set():
            try:
                worked = self.process_batch()
                self._heartbeat("running")
            except Exception:
                self._heartbeat("error")
                worked = False
            if not worked:
                self.stop_event.wait(settings.attachment_parse_poll_interval)

    def process_batch(self) -> bool:
        with self.store.connect() as conn:
            conn.execute("""UPDATE ingestion.worker_tasks SET status='pending',locked_by=NULL,locked_until=NULL,updated_at=now()
                WHERE task_type='attachment_parse' AND status='processing' AND locked_until < now()""")
            rows = conn.execute("""WITH picked AS (
                    SELECT t.id,t.payload->>'attachment_id' attachment_id FROM ingestion.worker_tasks t
                    JOIN ingestion.attachments a ON a.id=(t.payload->>'attachment_id')::bigint
                    WHERE t.task_type='attachment_parse' AND t.status IN ('pending','failed') AND t.next_run_at<=now()
                      AND a.download_status='completed' AND a.parse_status='pending' AND a.storage_key IS NOT NULL
                    ORDER BY t.id LIMIT %s FOR UPDATE SKIP LOCKED)
                UPDATE ingestion.worker_tasks t SET status='processing',attempts=attempts+1,locked_by=%s,
                    locked_until=now()+interval '30 minutes',updated_at=now() FROM picked
                WHERE t.id=picked.id RETURNING t.id,picked.attachment_id""",
                (settings.attachment_parse_batch_size, self.worker_id)).fetchall()
        for task_id, raw_attachment_id in rows:
            attachment_id, document_id = int(raw_attachment_id), None
            try:
                context = self.store.get_attachment_context(attachment_id)
                if context is None:
                    raise RuntimeError("attachment not found")
                self.store.begin_attachment_parse(attachment_id)
                if self.store.current_attachment_document(context) is not None:
                    self.store.skip_attachment_parse(attachment_id, int(task_id))
                    continue
                stale_ids = self.store.stale_attachment_document_ids(attachment_id)
                for stale_id in stale_ids:
                    delete_document_chunks(stale_id)
                self.store.delete_documents(stale_ids)
                result = self.parser.parse(context)
                metadata = {"attachment_id": attachment_id, "file_name": context.get("file_name"),
                            "extension": context.get("extension"), "mime_type": context.get("mime_type"),
                            "source_content_hash": context.get("content_hash"), "parser": result.parser,
                            "parser_version": result.parser_version, "derived_markdown_key": result.derived_markdown_key,
                            "canonical_key": result.canonical_key, "asset_keys": result.asset_keys, "headings": result.headings}
                document_id = self.store.upsert_attachment_document(context, result.content, metadata)
                self.store.finish_attachment_parse(attachment_id, int(task_id), document_id)
            except Exception as exc:
                if document_id is not None:
                    self.store.mark_failed(document_id, str(exc))
                self.store.fail_attachment_parse(attachment_id, int(task_id), str(exc))
        return bool(rows)

    def _heartbeat(self, status: str) -> None:
        try:
            with self.store.connect() as conn:
                conn.execute("""INSERT INTO ingestion.worker_runs(name,status,last_heartbeat,updated_at)
                    VALUES('rag-attachment-parse',%s,now(),now())
                    ON CONFLICT(name) DO UPDATE SET status=EXCLUDED.status,last_heartbeat=now(),updated_at=now()""", (status,))
        except Exception:
            pass

File: services/collectors/feishu/attachment_worker.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"strconv"

	"github.com/jackc/pgx/v5/pgxpool"
	"info-agent/feishu-collector/storage"
)

type FeishuAttachmentWorker struct {
	Pool        *pgxpool.Pool
	Adapter     FeishuAttachmentAdapter
	Store       *storage.Store
	TempDir     string
	WorkerID    string
	Account     string
	MaxAttempts int
}

func (w *FeishuAttachmentWorker) Process(ctx context.Context, taskID, attachmentID int64) error {
	var userID, conversationID, messageID, name, ext string
	var sourceMessageID string
	err := w.Pool.QueryRow(ctx, `SELECT sa.internal_account_id::text,c.id::text,m.id,m.source_message_id,COALESCE(a.file_name,''),COALESCE(a.extension,'')
        FROM ingestion.attachments a JOIN ingestion.messages m ON m.id=a.message_id
        JOIN ingestion.source_accounts sa ON sa.id=a.source_account_id JOIN ingestion.conversations c ON c.id=m.conversation_id
        WHERE a.id=$1 AND a.platform='feishu'`, attachmentID).Scan(&userID, &conversationID, &messageID, &sourceMessageID, &name, &ext)
	if err != nil {
		return fmt.Errorf("load attachment: %w", err)
	}
	if ext == "" || name == "" {
		return fmt.Errorf("attachment has no reliable file name or extension")
	}
	if err := w.Pool.QueryRow(ctx, `UPDATE ingestion.attachments SET download_status='downloading',last_error=NULL,updated_at=now() WHERE id=$1 AND download_status IN ('not_started','failed') RETURNING id`, attachmentID).Scan(&attachmentID); err != nil {
		return err
	}
	body, info, err := w.Adapter.DownloadByID(ctx, attachmentID)
	if err != nil {
		return err
	}
	defer body.Close()
	if w.TempDir == "" {
		w.TempDir = os.TempDir()
	}
	dir := filepath.Join(w.TempDir, "info-agent-feishu", strconv.FormatInt(attachmentID, 10))
	if err := os.MkdirAll(dir, 0o700); err != nil {
		return err
	}
	tmp := filepath.Join(dir, "payload.bin")
	f, err := os.OpenFile(tmp, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, 0o600)
	if err != nil {
		return err
	}
	_, copyErr := io.Copy(f, body)
	closeErr := f.Close()
	if copyErr != nil {
		return copyErr
	}
	if closeErr != nil {
		return closeErr
	}
	if info.Extension != "" {
		ext = info.Extension
	}
	if info.Name != "" {
		name = info.Name
	}
	// Use the PostgreSQL message id in the object key. The external Feishu
	// message id remains a separate field and is never used as PG identity.
	key := storage.ObjectKey(userID, "feishu", conversationID, messageID, strconv.FormatInt(attachmentID, 10), name)
	result, err := w.Store.PutFile(ctx, tmp, key, ext, info.ContentType)
	if err != nil {
		return err
	}
	_, err = w.Pool.Exec(ctx, `UPDATE ingestion.attachments SET file_size=$1,content_hash=$2,storage_provider='minio',storage_bucket=$3,storage_key=$4,download_status='completed',parse_status='pending',last_error=NULL,updated_at=now() WHERE id=$5`, result.Size, result.ContentHash, result.Bucket, result.Key, attachmentID)
	if err != nil {
		return err
	}
	_, err = w.Pool.Exec(ctx, `UPDATE ingestion.worker_tasks SET status='completed',locked_by=NULL,locked_until=NULL,completed_at=now(),last_error=NULL,updated_at=now() WHERE id=$1`, taskID)
	if err == nil {
		_, err = w.Pool.Exec(ctx, `INSERT INTO ingestion.worker_tasks(task_type,entity_id,payload) VALUES('attachment_parse',$1,$2::jsonb) ON CONFLICT(task_type,entity_id) DO UPDATE SET payload=EXCLUDED.payload,status='pending',next_run_at=now(),locked_by=NULL,locked_until=NULL,last_error=NULL,completed_at=NULL,updated_at=now()`, "attachment:"+strconv.FormatInt(attachmentID, 10), json.RawMessage(fmt.Sprintf(`{"platform":"feishu","attachment_id":%d}`, attachmentID)))
	}
	if err == nil {
		_ = os.RemoveAll(dir)
	}
	return err
}

func (w *FeishuAttachmentWorker) Fail(ctx context.Context, taskID, attachmentID int64, cause error) {
	message := cause.Error()
	if len(message) > 4000 {
		message = message[:4000]
	}
	_, _ = w.Pool.Exec(ctx, `UPDATE ingestion.attachments SET download_status='failed',last_error=$1,updated_at=now() WHERE id=$2`, message, attachmentID)
	_, _ = w.Pool.Exec(ctx, `UPDATE ingestion.worker_tasks SET status=CASE WHEN attempts>=max_attempts THEN 'dead' ELSE 'pending' END,next_run_at=now()+(LEAST(3600,60*(2^LEAST(attempts-1,6)))*interval '1 second'),locked_by=NULL,locked_until=NULL,last_error=$1,updated_at=now() WHERE id=$2`, message, taskID)
}

func (w *FeishuAttachmentWorker) RunOnce(ctx context.Context, limit int) (int, error) {
	rows, err := w.Pool.Query(ctx, `WITH picked AS (
        SELECT t.id,t.entity_id FROM ingestion.worker_tasks t
        JOIN ingestion.attachments a ON a.id=(t.payload->>'attachment_id')::bigint
        JOIN ingestion.source_accounts sa ON sa.id=a.source_account_id
        WHERE t.task_type='attachment_download' AND t.status='pending' AND t.next_run_at<=now()
          AND t.payload->>'platform'='feishu' AND sa.platform='feishu' AND sa.external_account_id=$1
        ORDER BY t.id LIMIT $2 FOR UPDATE OF t SKIP LOCKED)
        UPDATE ingestion.worker_tasks t SET status='processing',attempts=attempts+1,locked_by=$3,
        locked_until=now()+interval '10 minutes',updated_at=now() FROM picked WHERE t.id=picked.id
        RETURNING t.id,t.entity_id`, externalAccountID(w.Account), limit, w.WorkerID)
	if err != nil {
		return 0, err
	}
	defer rows.Close()
	count := 0
	for rows.Next() {
		var taskID int64
		var entity string
		if err := rows.Scan(&taskID, &entity); err != nil {
			continue
		}
		var attachmentID int64
		if _, err := fmt.Sscanf(entity, "attachment:%d", &attachmentID); err != nil {
			w.Fail(ctx, taskID, 0, fmt.Errorf("invalid attachment task entity"))
			continue
		}
		if err := w.Process(ctx, taskID, attachmentID); err != nil {
			w.Fail(ctx, taskID, attachmentID, err)
		}
		count++
	}
	return count, rows.Err()
}
